ComponentKit/StatefulViews/CKStatefulViewReusePool.mm

```diff
@@ -10,40 +10,76 @@
 #import "CKStatefulViewComponentController.h"
 
 #import "CKStatefulViewReusePool.h"
+#import "CKAssert.h"
 
 #import <unordered_map>
 
+struct FBStatefulReusePoolItemEntry {
+  UIView *view;
+  CKStatefulViewReusePoolPendingMayRelinquishBlock block;
+};
+
 class FBStatefulReusePoolItem {
 public:
-  FBStatefulReusePoolItem()
-  : views([NSMutableArray array]) {};
-
   UIView *viewWithPreferredSuperview(UIView *preferredSuperview)
   {
-    if ([views count] == 0) {
+    if (_entries.empty()) {
+      return nil;
+    }
+    // Preferentially return the parent view.
+    auto preferIt = std::find_if(_entries.begin(), _entries.end(),
+                           [preferredSuperview](const FBStatefulReusePoolItemEntry entry)->bool {
+                             return entry.view == preferredSuperview;
+                           });
+    if (preferIt != _entries.end()) {
+      FBStatefulReusePoolItemEntry entry = *preferIt;
+      _entries.erase(preferIt);
+      if (entry.block == NULL || entry.block()) {
+        return entry.view;
+      }
+    }
+
+    if (_entries.empty()) {
```

Comment:
I'd include a comment explaining why we're checking if `_entries` is empty again. Even though there's a comment explaining what's going on just a bit further down.

At first I thought this was a mistake.